src/lib/services/event-service.ts

```diff
@@ -35,4 +41,44 @@ export default class EventService {
             totalEvents,
         };
     }
+
+    async storeEvent(event: IBaseEvent): Promise<void> {
+        if (event.featureName && !event.tags) {
+            const tags = await this.featureTagStore.getAllTagsForFeature(
+                event.featureName,
+            );
+            return this.eventStore.store({ ...event, tags });
+        }
+        return this.eventStore.store(event);
+    }
+
+    async storeEvents(events: IBaseEvent[]): Promise<void> {
+        const featureNames: string[] = events
+            .map((event) => event.featureName)
+            .filter(
+                (featureName): featureName is string =>
+                    featureName !== undefined,
+            );
+
+        if (featureNames.length) {
+            const tags = await this.featureTagStore.getAllByFeatures(
+                featureNames,
+            );
+
+            const eventsWithTags = events.map((event) => ({
+                ...event,
+                tags:
+                    event.tags ||
+                    tags
+                        .filter((tag) => tag.featureName === event.featureName)
+                        .map((tag) => ({
+                            value: tag.tagValue,
+                            type: tag.tagType,
+                        })),
+            }));
+
+            return this.eventStore.batchStore(eventsWithTags);
```

Comment:
This code might be better if we process it once and convert it to a `Map<string, string[]>` with the key being the feature name and the value the list of tags.

We're also iterating the list of events twice and I think you can do it in one go. I'd do something like:

```typescript
for (const enhancer of enhancers) {
    events = await enhance(events);
}
return this.eventStore.batchStore(events);
```
where enhance can be "add tags" or decorate with username (might be handy for one of our next tasks in Q4)

Reply:
What do you think of https://github.com/Unleash/unleash/pull/4830/commits/03976efb1815b4f55d550650af24347e1bcad9f0?

Alternatively, we could also fetch all tags, build a Set out of them, and only then map them. Something like this:

```ts
const featureTagsMap: Map<string, ITag[]> = new Map();
const allTags = await this.featureTagStore.getAll();

for (const tag of allTags) {
  const featureTags = featureTagsMap.get(tag.featureName) || [];
  featureTags.push({ value: tag.tagValue, type: tag.tagType });
  featureTagsMap.set(tag.featureName, featureTags);
}
```

But if we consider the total amount of events we're storing vs total amount of tags we could be fetching, I think the approach in the commit should be more performant, along with the advantage of being more readable. But I'm open to suggestions.

Reply:
I love it, I'm thinking about your alternative or just keeping the previous implementation (because now that I look at the new code I'm thinking would it be faster to iterate twice a list in memory or to query the DB multiple times per each feature? I believe it depends on the type of event, but in most cases, 2 loops in memory will be faster). In any case, I think it's fine to go with this option because it's currently the one we have in the code (before each event we query the DB), but also, feel free to change it

Reply:
True, was questioning this myself as well. What do you think of something like this instead?

```ts
private async enhanceEventsWithTags(
    events: IBaseEvent[],
): Promise<IBaseEvent[]> {
    const featureNamesSet = new Set<string>();
    for (const event of events) {
        if (event.featureName && !event.tags) {
            featureNamesSet.add(event.featureName);
        }
    }

    const featureTagsMap: Map<string, ITag[]> = new Map();
    const allTagsInFeatures = await this.featureTagStore.getAllByFeatures(
        Array.from(featureNamesSet),
    );

    for (const tag of allTagsInFeatures) {
        const featureTags = featureTagsMap.get(tag.featureName) || [];
        featureTags.push({ value: tag.tagValue, type: tag.tagType });
        featureTagsMap.set(tag.featureName, featureTags);
    }

    for (const event of events) {
        if (event.featureName && !event.tags) {
            event.tags = featureTagsMap.get(event.featureName);
        }
    }

    return events;
}
```

Reply:
I like this approach :+1:

Reply:
Thanks, addressed in https://github.com/Unleash/unleash/pull/4830/commits/c2cbad3063f9768ce1b29ac60e5935769fcc6f5f